yarl/_url.py

```diff
@@ -1116,7 +1122,59 @@ def join(self, url):
         # See docs for urllib.parse.urljoin
         if not isinstance(url, URL):
             raise TypeError("url should be URL")
-        return URL(urljoin(str(self), str(url)), encoded=True)
+        other: URL = url
+        scheme = other.scheme or self.scheme
+        parts = {
+            k[4:]: getattr(self, k) or ""
+            for k in ("raw_authority", "raw_path", "raw_query_string", "raw_fragment")
+        }
+        parts["scheme"] = scheme
+
+        if scheme != self.scheme or scheme not in uses_relative:
+            return URL(str(other))
+
+        # scheme is in uses_authority as uses_authority is a superset of uses_relative
+        if scheme in uses_authority and other.authority:
+            parts.update(
+                {
+                    k[4:]: getattr(other, k) or ""
+                    for k in (
+                        "raw_authority",
+                        "raw_path",
+                        "raw_query_string",
+                        "raw_fragment",
+                    )
+                }
+            )
```

Comment:
This is getting rather long, it probably more readable to write it out in both places

Reply:
I propose a function instead